src/magic_folder/downloader.py

```diff
@@ -209,15 +208,15 @@ class IMagicFolderFilesystem(Interface):
     there are no 'partial' files in the magic-folder.
     """
 
-    def download_content_to_staging(remote_snapshot, tahoe_client):
+    def download_content_to_staging(relpath, remote_snapshot, tahoe_client):
```

Comment:
Removing the tie between the snapshot and its name seems like a bad move. Why can't the snapshot know its own name?

Reply:
It is generally better to have a single source of truth, and given that we *need* to have the name in the directory entry, it seems preferable to have that be the source of truth. Also, I can imagine a future where we detect renames, in which case, we may end up writing a file to a renamed path.

At this *particular* point, I think it makes sense for `IMagicFolderFilesystem` to not know about snapshots, this is an incremental step towards that. (That is, we should also be passing in the capabilitiy, mtime and author, to the three functions, rather than accessing those attributes on the passed in snapshot).

Reply:
What are you saying is "the single source of truth"?

Reply:
The (demangled) name from the DMD that we discovered this snapshot via.

Reply:
Part of the reason that I prefer using that name, is that there can't be collisions by construction[1], but the data format allows for clashes in the metadata name (even if our current code never constructs such a thing).

[1] Unless or mangling isn't unique. Or, if tahoe returned a JSON object with duplicate keys, but that would be a violation of *tahoe*s data model

Reply:
So why can't the snapshot know its own name?

Reply:
Do you mean the python `RemoteSnapshot` object?  I don't have any objection major[1] objection to it knowing the name it was access by, though it would require some refactoring to that, since we don't currently have the name where we are constructing the snapshot . Also, for snapshots that we discover as parents, the tie to the name of the child is less strong (though in this case, I don't think we ever *use* the name).

On the other hand, I don't see any clear benefit to having it store the name either. One downside is that a `RemoteSnapshot` would then no longer be a python representation of snapshot capability (this is why it would no longer be a great representation of parent snapshots, as mentioned above).

[1] Though it might be confusing to have both a name, and a potentially different name in the metadata.

Reply:
>  I don't have any objection major objection [...]

To be clear, I don't think we should make `RemoteSnapshot.name` be the name from the DMD. I think it would make the code (in particular `RemoteSnapshotCacheService`) more complicated, and the make the correspondence between the persistent (i.e. on grid) data representation and the in-process one more tenuous. The only advantage I can see is not needing to pass an extra value a few places, which isn't clearly better.

Reply:
`RemoteSnapshot.name` _is_ the name from the DMD.

Reply:
> `RemoteSnapshot.name` _is_ the name from the DMD.

No: https://github.com/LeastAuthority/magic-folder/blob/6f8c379e253effd0c5144f7f4a1170acd93fa34c/src/magic_folder/snapshot.py#L428-L443 It is the name from the metadata blob in the snapshot. Given that `create_snapshot_from_capability` only takes the capability, it *can't* have the name from the DMD that the points at that capability.

(Granted, it is currently always the same, but as I argued above, we should get the value from the source of truth, which is not where it comes `RemoteSnapshot.name` comes from.)

Reply:
> Granted, it is currently always the same

Yes, that's what I mean. "A snapshot's name" is its entry in the DMD.

Reply:
My point is we should program defensively against those getting out of sync, either due to a bug, or because in the future we (e.g.) detect renames. So, **even though they are currently the same**, we should consistently use only one of the values, rather than sometimes using one and sometimes using the other.

And, before you point out we could bump the version if we want to decouple them, if we program defensively now, we may not need to do that.

Reply:
I'm talking about all the places that take an extra "relpath" arg in this PR beside (or sometimes instead of) a `RemoteSnapshot` .. that makes it easy to pass the wrong thing instead of the snapshot knowing its name (and/or relpath).

Reply:
> I'm talking about all the places that take an extra "relpath" arg in this PR [...] that makes it easy to pass the wrong thing

I do think that is a reasonable concern. As it is, we don't have an object that bundles a remote snapshot and the name in the DMD that we read it from (if we are programming defensively, `RemoteSnapshot` *is not that*). I don't think it worth it to introduce one here, particularly as I want most of the places this would affect to stop taking snapshots at all.

For me, it comes down to being resilient against inconsistencies or changes in format of our persistent data (i.e. what we read from the tahoe grid) outweights the risk of passing the wrong arguments. The later is *entirely* under our control and doesn't impose any constraints on what we do in the future. The way we interpret data on-grid *does* constrain how we can evolve our software in the future, so I am much more concerned about that.